sdk/src/main/java/io/opentelemetry/sdk/trace/RecordEventsSpanImpl.java

```diff
@@ -0,0 +1,743 @@
+    }
+    return events;
+  }
+
+  @GuardedBy("this")
+  private TraceEvents<Link> getInitializedLinks() {
+    if (links == null) {
+      links = new TraceEvents<Link>(traceConfig.getMaxNumberOfLinks());
+    }
+    return links;
+  }
+
+  @GuardedBy("this")
+  private Status getStatusWithDefault() {
+    return status == null ? Status.OK : status;
+  }
+
+  @Override
+  @Nullable
+  public RecordEventsSpanImpl getNext() {
+    return next;
+  }
+
+  @Override
+  public void setNext(@Nullable RecordEventsSpanImpl element) {
+    next = element;
+  }
+
+  @Override
+  @Nullable
+  public RecordEventsSpanImpl getPrev() {
+    return prev;
+  }
+
+  @Override
+  public void setPrev(@Nullable RecordEventsSpanImpl element) {
+    prev = element;
+  }
+
+  /**
+   * Interface to handle the start and end operations for a {@link Span} only when {@link
+   * #isRecordingEvents()} is true.
+   *
+   * <p>Implementation must avoid high overhead work in any of the methods because the code is
+   * executed on the critical path.
+   *
+   * <p>One instance can be called by multiple threads in the same time, so the implementation must
+   * be thread-safe.
+   */
+  public interface StartEndHandler {
+    void onStart(RecordEventsSpanImpl span);
+
+    void onEnd(RecordEventsSpanImpl span);
+  }
+
+  // A map implementation with a fixed capacity that drops events when the map gets full. Eviction
+  // is based on the access order.
+  private static final class AttributesWithCapacity extends LinkedHashMap<String, AttributeValue> {
+    private final long capacity;
+    private int totalRecordedAttributes = 0;
+    // Here because -Werror complains about this: [serial] serializable class AttributesWithCapacity
+    // has no definition of serialVersionUID. This class shouldn't be serialized.
+    private static final long serialVersionUID = 42L;
+
+    private AttributesWithCapacity(long capacity) {
+      // Capacity of the map is capacity + 1 to avoid resizing because removeEldestEntry is invoked
+      // by put and putAll after inserting a new entry into the map. The loadFactor is set to 1
+      // to avoid resizing because. The accessOrder is set to true.
+      super((int) capacity + 1, 1, /*accessOrder=*/ true);
+      this.capacity = capacity;
+    }
+
+    // Users must call this method instead of put to keep count of the total number of entries
+    // inserted.
+    private void putAttribute(String key, AttributeValue value) {
+      totalRecordedAttributes += 1;
+      put(key, value);
+    }
+
+    private int getNumberOfDroppedAttributes() {
+      return totalRecordedAttributes - size();
+    }
+
+    // It is called after each put or putAll call in order to determine if the eldest inserted
+    // entry should be removed or not.
+    @Override
+    protected boolean removeEldestEntry(Map.Entry<String, AttributeValue> eldest) {
+      return size() > this.capacity;
+    }
+  }
+
+  private static final class TraceEvents<T> {
```

Comment:
Consider to avoid an allocation and embed this directly into the Span:

```java
private int totalRecordedEvents;
private EvictingQueue<Events> events;
private int totalRecordedLinks;
private EvictingQueue<Link> links;
```

Reply:
SG, updated.